lib/elixir/test/elixir/uri_test.exs

```diff
@@ -225,4 +225,52 @@ defmodule URITest do
     assert URI.to_string(URI.parse("http://google.com")) == "http://google.com"
     assert URI.to_string(URI.parse("//user:password@google.com/")) == "//user:password@google.com/"
   end
+
+  test "merge/2" do
+    assert_raise ArgumentError, fn ->
+      URI.merge("/relative", "")
+    end
+
+    assert URI.merge("http://google.com/foo", "http://example.com/baz") |> to_string == "http://example.com/baz"
+
+    assert URI.merge("http://example.com", URI.parse("/foo")) |> to_string == "http://example.com/foo"
+
+    base = URI.parse("http://example.com/foo/bar")
+    assert URI.merge(base, "") |> to_string == "http://example.com/foo/bar"
+    assert URI.merge(base, "#fragment") |> to_string == "http://example.com/foo/bar#fragment"
+    assert URI.merge(base, "?query") |> to_string == "http://example.com/foo/bar?query"
+    assert URI.merge(base, %URI{path: ""}) |> to_string == "http://example.com/foo/bar"
+    assert URI.merge(base, %URI{path: "", fragment: "fragment"}) |> to_string == "http://example.com/foo/bar#fragment"
+
+    base = URI.parse("http://example.com")
+    assert URI.merge(base, "/foo") |> to_string == "http://example.com/foo"
+    assert URI.merge(base, "foo") |> to_string == "http://example.com/foo"
+
+    base = URI.parse("http://example.com/foo/bar")
+    assert URI.merge(base, "/baz") |> to_string == "http://example.com/baz"
+    assert URI.merge(base, "baz") |> to_string == "http://example.com/foo/baz"
+    assert URI.merge(base, "../baz") |> to_string == "http://example.com/baz"
```

Comment:
Or does the spec say to always discard the last segment? If so, we should also include tests for `URI.parse("http://example.com/foo/bar/")`.